### ISSUE TITLE
ESP32 support?

### ISSUE THREAD

**Issue by ligius-** [maaaaz/openocd-windows-buspirate#1]
I've tried to compile OpenOCD myself for win64 with buspirate and esp32 support:

`/configure --enable-buspirate --enable-maintainer-mode --build=i686-pc-cygwin --host=x86_64-w64-mingw32`

Result:
`configure: error: buspirate currently not supported by MinGW32 hosts`

However it refuses to compile with the given host argument, not sure how to go from there. Without the host argument it only runs in Cygwin. Your build runs fine, but it lacks ESP32 support:
https://github.com/espressif/openocd-esp32

Is there any way you can rebuild this with ESP32 included or give me some hints on how to do it myself?

**Comment by maaaaz** [comment]
Hello,

- First thing, I didn't use MinGW but the Cygwin GCC to compile my version
- Then, try to look at [this](http://esp-idf.readthedocs.io/en/latest/api-guides/jtag-debugging/building-openocd-windows.html) and [this](http://esp-idf.readthedocs.io/en/latest/api-guides/jtag-debugging/setup-openocd-windows.html) for ESP32


Cheers.

**Comment by ligius-** [comment]
I've edited my initial comment a few minutes after it was added, "MinGW" was changed to Cygwin. I'm looking specifically for the build and host options and whether some patches were applied. I know that at least I had to apply a patch to buspirate.c by Michal Demin.

**Comment by maaaaz** [comment]
"Without the host argument it only runs in Cygwin."
* So you managed to compile something but it can only run inside Cygwin while my version can run without, it that the issue ?


"Your build runs fine, but it lacks ESP32 support: https://github.com/espressif/openocd-esp32"
* So you would me to compile that espressif version of OpenOCD ?

**Comment by ligius-** [comment]
Thank you for the reply.
I will go into more details:
My problem was that the ESP32 custom OpenOCD build did not include buspirate nor bitbang_remote. Your build does not include ESP32.

The "runs only under Cygwin issue" I was having turned out to be a PEBKAC: I didn't have the needed DLLs next to the executable, but no error message was generated when launching the executable. Fixed by copying Cygwin1.dll and the others into the launch path.

Integrating ESP32 is not trivial but not very hard, I've done it on the latest OpenOCD build. I made a patch file if anyone is interested, but it's untested, because:

I cannot get OpenOCD + buspirate to compile for Windows. First of all it's designed to throw an error in the configure. After applying an older patch to buspirate.c, the compiler complains about missing [Windows] types or missing termios.h. That's as far as I could get.

My request, if possible, could you provide a description of what changes you did? For example, buspirate support is disabled by this: http://openocd.zylin.com/gitweb?p=openocd.git;a=blob;f=configure.ac;h=562ec5a7cc47a1cc4bafcf5f48992bd483368943;hb=HEAD#l431
Can you also describe which environment (Cygwin? MSYS?) are you using and any other build/configure parameters?

I can only hope that ESP32 will be pushed upstream in the future, will contact them.

**Comment by maaaaz** [comment]
I managed to compile the espressif version in Cygwin with the same steps described in my version README. I haven't edited any file:
```
$ git clone https://github.com/espressif/openocd-esp32.git && cd openocd-esp32
$ ./bootstrap && ./configure --enable-buspirate --enable-maintainer-mode
$ make && make install
```

You can find the build [openocd-0.10.0-Win32-with_buspirate_and_esp32_support.zip](https://github.com/maaaaz/openocd-windows-buspirate/files/1434576/openocd-0.10.0-Win32-with_buspirate_and_esp32_support.zip), test it and tell me if it works well:
```
> .\openocd.bat -c "target types"
Open On-Chip Debugger 0.10.0-dev-g7aeadf8e (2017-11-01-15:52)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
arm7tdmi arm9tdmi arm920t arm720t arm966e arm946e arm926ejs fa526 feroceon dragonite xscale cortex_m cortex_a cortex_r4 arm11 ls1_sap mips_m4k avr dsp563xx dsp5680xx testee avr32_ap7k hla_target nds32_v2 nds32_v3 nds32_v3m esp108 esp32 or1k quark_x10xx quark_d20xx
```
For your information, you indeed have to copy Cygwin DLLs in the same directory if you want to be able to run the program without a Cygwin environment. I use that command to automate it:
```
$ ldd openocd.exe | grep -iv WINDOWS | cut -d '>' -f 2 | cut -d ' ' -f 2-|sed -r 's/(.*)\s[(].*/\1/g'| xargs -n1 -I {} sh -c 'cp "{}" .'
```

Cheers.

**Comment by ligius-** [comment]
Thank you, your build works well. I tried the same steps and I could compile at as well.
It seems that my issues were: --host and --build do not need to be specified, even though the platform triplet contains "unknown", it will still build fine for w64. The patch I applied to buspirate.c broke the build, so, restoring the original file got the build back working.

I added --enable-remote_bitbang to my configure line, this would allow debugging via ESP8266, though not sure if it's faster than BusPirate.

Thank you again for your support and time.
